base/compiler/typeutils.jl

```diff
@@ -132,7 +132,7 @@ function valid_as_lattice(@nospecialize(x))
 end
 
 function valid_typeof_tparam(@nospecialize(t))
-    if t === Symbol || isbitstype(t)
+    if t === Symbol || t === Module || isbitstype(t)
         return true
     end
     isconcretetype(t) || return false
```

Comment:
This checks if `t` is a DataType, but then may unwrap a NamedTuple, which must return a DataType too, but we don't have a typeassert there so the compiler also knows that. Can you add a typeassert there?

Aside: we have this definition:
```julia
valid_tparam_type(U::UnionAll) = valid_tparam_type(unwrap_unionall(U))
```
But I cannot think of how that would ever return true, since by definition, the content of the UnionAll must be an incomplete type (has_free_typevars), and thus will not be concrete, and thus will also not be any of the concrete types listed just above (Symbol, Module, or isbitstype)

Reply:
Yeah, I think you're right, I think we can just do 

```
valid_tparam_type(U::UnionAll) = false
```

@oscardssmith can you try that and run pkgeval?

Reply:
I'll do that in a separate PR.